gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,359 @@
+
+        cloud_event = CloudEvent(
+            id=str(uuid.uuid4()),
+            source=f"{settings.gateway_url}qod-provisioning/v0.2/device-qos/{id}",
+            type=Type.org_camaraproject_quality_on_demand_v1_qos_status_changed,
+            specversion=Specversion.field_1_0,
+            datacontenttype=Datacontenttype.application_json,
+            data=sub_info.model_dump(mode="json", exclude_unset=True),
+            time=datetime.datetime.now(),
+        )
+
+        sink = sub_info.sink
+        if sink is None:
+            raise ResourceNotFound()
+
+        await self.httpx_client_callback.post(
+            str(sink),
+            content=cloud_event.model_dump_json(exclude_unset=True, by_alias=True),
+        )
+
+        await self._save_subscription_info(id, sub_info)
+
+    def _get_qod_status(
+        self, nef_response: UserPlaneNotificationData
+    ) -> QosStatus | None:
+        status = None
+        for report in nef_response.eventReports:
+            if report.event == UserPlaneEvent.QOS_GUARANTEED:
+                status = QosStatus.AVAILABLE
+            elif report.event == UserPlaneEvent.FAILED_RESOURCES_ALLOCATION:
+                status = QosStatus.UNAVAILABLE
+            elif report.event == UserPlaneEvent.QOS_NOT_GUARANTEED:
+                status = QosStatus.REQUESTED
+            elif report.event == UserPlaneEvent.SUCCESSFUL_RESOURCES_ALLOCATION:
+                status = QosStatus.AVAILABLE
+        return status
+
+    async def _save_subscription_info(self, id: str, data: SessionInfo) -> None:
+        key = f"{_prefix_info}:{id}"
+
+        await self.redis.set(
+            key, data.model_dump_json(exclude_unset=True, by_alias=True)
+        )
+
+    def _get_flow_description(self, req: CreateSession | SessionInfo) -> List[str]:
+        flow_description = []
+        if req.applicationServerPorts is None or req.devicePorts is None:
+            raise ResourceNotFound()
+
+        if req.applicationServerPorts.ports is None or req.devicePorts.ports is None:
+            raise ResourceNotFound()
+
+        if req.device and req.device.ipv4Address is not None:
+            flow_description.append(
+                f"permit in ip from {req.device.ipv4Address.publicAddress} {req.devicePorts.ports[0]} to {req.applicationServer.ipv4Address} {req.applicationServerPorts.ports[0]}"
+            )
+            flow_description.append(
+                f"permit in ip from {req.applicationServer.ipv4Address} {req.applicationServerPorts.ports[0]} to {req.device.ipv4Address.publicAddress} {req.devicePorts.ports[0]}"
+            )
+
+        elif req.device and req.device.ipv6Address:
+            flow_description.append(
+                f"permit in ip from {req.device.ipv6Address.exploded} to {req.applicationServer.ipv6Address}"
+            )
+            flow_description.append(
+                f"permit in ip from {req.applicationServer.ipv6Address} to {req.device.ipv6Address.exploded}"
+            )
```

Comment:
It's not one or another you should have two flowInfos if both IPv4 and IPv6 are provided. You should also include ports for IPv6. And finally, you should put all ports in the description you can separate them with commas, and for ranges you can do `port-port`

You can for example have `23,80-443` in the same description.

Reply:
Forgot to add but this is defined in [rfc6733](https://www.rfc-editor.org/rfc/rfc6733) in IPFilterRule